pip/vcs/git.py

```diff
@@ -134,7 +134,7 @@ def obtain(self, dest):
             rev_options = [rev]
             rev_display = ' (to %s)' % rev
         else:
-            rev_options = ['origin/master']
+            rev_options = ['refs/remotes/origin/HEAD']
```

Comment:
@cjerdonek I'm not sure I agree. I think this is perhaps close to what you're suggesting:

```python
@patch('pip.vcs.call_subprocess')
@patch('pip.vcs.os.path.exists', lambda _: True)
@pytest.mark.network
def test_obtain_reset(call_subprocess_mock, script):
    git = Git('git+file:///somewhere')
    git.compare_urls = lambda *args: True
    git.check_version = lambda *args: False
    git.obtain(script.scratch_path / 'test')

    assert call_subprocess_mock.call_args_list[4][0][0] == [
        'git', 'reset', '--hard', '-q', 'refs/remotes/origin/HEAD'
    ]
```
This test is mostly just checking that `rev_options` has changed, and doesn't really motivate the fix I'm proposing. Further more, it feels pretty fragile -- if we're not actually installing a package in the test, we need to do some monkeypatching to make the `Git` instance think the package is already installed so that it attempts to `update` it, and then there's a slew of sequential subprocess calls, only one of which we really care about. 

I understand that such high-level tests are slow and expensive, but I think in this case it does make sense, although I'm open to further suggestions -- I'm relatively new to contributing to pip, so I'm not totally familiar with how we test it.